### PR TITLE
Replace CLAP with Essentia.js + ONNX instrument classification

### LINKED ISSUE #268: Replace CLAP with Essentia.js WASM + Discogs-EffNet + MTG-Jamendo instrument classification
## Problem

The CLAP model (`Xenova/larger_clap_music_and_speech`) classifies every isolated stem as "singing vocals" regardless of actual content. This has persisted through four iterations:

| PR | Approach | Result |
|---|---|---|
| #250 | Swap AST → CLAP zero-shot (`clap-htsat-unfused`) | Everything = vocals |
| #259 | Filename-based primary + CLAP fallback | Closed (not merged) |
| #261 | Extract loudest 10s segment before inference | Everything = vocals |
| #266 | Switch to music-trained variant (`larger_clap_music_and_speech`) | Everything = vocals |

See #249 for full history.

## Solution

Replace the CLAP zero-shot pipeline with a purpose-built instrument classification pipeline using **Essentia.js WASM** for feature extraction and **ONNX Runtime Web** for inference.

### Pipeline

```
Audio → mono → resample to 16 kHz
  → Essentia.js WASM: mel spectrogram (96 mel bands, frame=512, hop=256)
  → log compression: log10(1 + 10000 * mel_bands)
  → patch into [128, 96] batches (128 time frames × 96 mel bands)
  → discogs-effnet-bsdynamic-1.onnx (~18 MB) → 1280-dim embeddings
  → average embeddings across patches
  → mtg_jamendo_instrument-discogs-effnet-1.onnx (~2.7 MB) → 40 instrument predictions
  → map top prediction to InstrumentLabel
```

### Models

Both hosted at `essentia.upf.edu` with Cache API caching:

| Model | URL | Size | Input | Output |
|---|---|---|---|---|
| Discogs-EffNet | `essentia.upf.edu/models/feature-extractors/discogs-effnet/discogs-effnet-bsdynamic-1.onnx` | 18 MB | mel spectrogram `[n, 128, 96]` | embeddings `[n, 1280]` |
| MTG-Jamendo instrument head | `essentia.upf.edu/models/classification-heads/mtg_jamendo_instrument/mtg_jamendo_instrument-discogs-effnet-1.onnx` | 2.7 MB | embeddings `[1, 1280]` | 40 sigmoid predictions |

Total download: ~20.7 MB (vs 205 MB for CLAP music-trained).

### Mel spectrogram parameters (matching Essentia's TensorflowInputMusiCNN)

| Parameter | Value |
|---|---|
| sampleRate | 16,000 Hz |
| frameSize | 512 |
| hopSize | 256 |
| numberBands | 96 |
| weighting | linear |
| warpingFormula | slaneyMel |
| normalize | unit_tri |
| log compression | `log10(1 + 10000 * mel_bands)` |
| patch size | 128 frames (EffNet-specific; MusiCNN uses 187) |

### 40 Jamendo classes → 10 InstrumentLabel categories

```
vocals     ← voice
guitar     ← acousticguitar, classicalguitar, electricguitar, guitar
bass       ← acousticbassguitar, bass, doublebass
drums      ← drums, drummachine
keyboard   ← accordion, electricpiano, keyboard, organ, piano, pipeorgan, rhodes
strings    ← cello, harp, orchestra, strings, viola, violin
brass      ← brass, horn, trombone, trumpet
woodwind   ← clarinet, flute, harmonica, oboe, saxophone
synth      ← computer, pad, sampler, synthesizer
percussion ← beat, bell, bongo, percussion
```

### Essentia.js API

The `EssentiaTFInputExtractor` class with `extractorType='musicnn'` computes mel spectrograms with exactly the parameters EffNet needs (96 mel bands, frame=512, hop=256 at 16 kHz). The only difference is the patch size (MusiCNN=187, EffNet=128) — handle this by taking the raw mel frames from `computeFrameWise()` and re-patching into groups of 128.

Key API:
```js
// In worker: load WASM, create extractor
import EssentiaWASM from 'essentia.js/dist/essentia-wasm.es.js';
import { EssentiaTFInputExtractor } from 'essentia.js/dist/essentia.js-model.es.js';

const wasmModule = await EssentiaWASM();
const extractor = new EssentiaTFInputExtractor(wasmModule, 'musicnn');

// Compute mel spectrogram frames for whole audio signal
const features = extractor.computeFrameWise(monoAudio16kHz);
// features.melSpectrum is a flat Float32Array of all frames × 96 bands
// features.melBandsSize = 96, features.patchSize = 187 (ignore, use 128)
```

### 1. Update `src/services/instrumentLabels.ts`

Replace CLAP candidate labels with Jamendo-to-InstrumentLabel mapping:
- Remove `CANDIDATE_LABELS` array and `CANDIDATE_TO_INSTRUMENT` map
- Add `JAMENDO_CLASSES` array (40 strings, ordered to match model output)
- Add `JAMENDO_TO_INSTRUMENT` mapping (40 → 10 categories as above)
- Update `mapToInstrumentLabel()` to use new mapping
- Keep `InstrumentLabel` type and `FALLBACK_LABEL` unchanged

### 2. Create `src/services/melSpectrogram.ts`

New module for essentia.js WASM mel spectrogram computation:
- Wrap `EssentiaTFInputExtractor` with `musicnn` type
- Export `computeMelSpectrogram(monoAudio: Float32Array): Float32Array[]` that returns 128-frame patches of 96 mel bands
- Handle WASM module initialization (lazy, once)
- Apply log compression: `log10(1 + 10000 * x)`

### 3. Rewrite `src/services/classification.worker.ts`

Replace CLAP/Transformers.js pipeline with essentia.js + ONNX:
- Import essentia.js WASM ES module and `EssentiaTFInputExtractor`
- Import `onnxruntime-web` for ONNX inference
- Load two ONNX models with Cache API caching (adapt `ModelCache.ts` pattern)
- Preprocessing: `downmixToMono` → `resample` to 16 kHz → `computeMelSpectrogram` → ONNX EffNet → ONNX instrument head
- Keep the same message protocol (`ClassifyRequest`, `ClassifyResponse`, `DownloadProgressMessage`)
- Keep download progress tracking (implement via `fetch` with `content-length` header)
- Remove `@huggingface/transformers` import
- Remove `extractLoudestSegment` usage (EffNet processes full mel spectrogram via patches, no truncation needed)

### 4. Update `src/services/InstrumentClassificationService.ts`

- Update `MODEL_SAMPLE_RATE` from 48,000 to 16,000
- Update main-thread fallback to use essentia.js + ONNX instead of `@huggingface/transformers`
- Remove `CANDIDATE_LABELS` import (no longer needed — model has fixed output classes)
- Keep all signal ownership, caching, worker lifecycle, and public API unchanged

### 5. Update `src/services/ModelCache.ts`

- May need a different cache name (e.g. `'essentia-models'`) since models come from `essentia.upf.edu` not HuggingFace
- Same stale-while-revalidate pattern applies

### 6. Update tests

**`src/services/__tests__/instrumentLabels.test.ts`:**
- Test new `JAMENDO_CLASSES` array (40 entries)
- Test all 40 → 10 mappings
- Test fallback to 'unknown'

**`src/services/__tests__/melSpectrogram.test.ts`** (new):
- Test mel spectrogram computation with mock essentia.js WASM
- Test patching into 128-frame groups
- Test log compression

**`src/services/__tests__/InstrumentClassificationService.test.ts`:**
- Update mock from `@huggingface/transformers` to `onnxruntime-web`
- Update worker message protocol tests if needed
- Keep all signal/caching/lifecycle tests (they don't depend on the model)
- Update main-thread fallback tests

**`src/services/__tests__/classification.worker.test.ts`** (if exists):
- Update for new pipeline

### 7. Cleanup

- Remove `src/services/audioSegment.ts` and its tests (loudest segment extraction no longer needed)
- Remove `@huggingface/transformers` from `package.json` if no longer used elsewhere
- Run `npm test`, `npm run lint`, `npm run build`

## Dependencies

**Add:**
- `essentia.js` (^0.1.3) — WASM backend for mel spectrogram computation (~2.7 MB packed)
- `onnxruntime-web` (^1.24.3) — ONNX model inference (already a transitive dep, making it direct)

**Potentially remove:**
- `@huggingface/transformers` — check if used elsewhere before removing

## Why this should work better than CLAP

- **Purpose-built**: MTG-Jamendo instrument model trained specifically on instrument classification (40 classes, 0.78 ROC-AUC on test set)
- **No zero-shot bias**: Fixed label set from supervised training — no text embedding similarity issues
- **Smaller download**: ~20.7 MB total vs 205 MB for CLAP music-trained
- **Proven pipeline**: Essentia's Discogs-EffNet + classification heads are widely used in MIR research

## Branch

Develop on `claude/fix-instrument-classification-MXwK3`. Both `essentia.js` and `onnxruntime-web` are already installed on this branch.

### PR DESCRIPTION
**Pull request by vsandvold**
## Summary

- Replace CLAP zero-shot pipeline (which classified everything as vocals) with a purpose-built instrument classification pipeline using **Essentia.js WASM** for mel spectrogram extraction and **ONNX Runtime Web** for inference
- New pipeline: mono 16 kHz audio → mel spectrogram (96 bands) → Discogs-EffNet embeddings (1280-dim) → MTG-Jamendo instrument head (40 classes → 10 categories)
- Total model download ~20.7 MB (vs 205 MB for CLAP), with Cache API caching

### Changes

- **`instrumentLabels.ts`**: Replace 10 CLAP candidate labels with 40 Jamendo class → 10 InstrumentLabel mapping
- **`melSpectrogram.ts`** (new): Essentia.js WASM mel spectrogram computation with 128-frame patching and log compression
- **`classification.worker.ts`**: Rewrite for Essentia.js + ONNX Runtime pipeline (EffNet + Jamendo instrument head)
- **`InstrumentClassificationService.ts`**: Update main-thread fallback to use ONNX pipeline; change target sample rate from 48 kHz to 16 kHz
- **`ModelCache.ts`**: Rewrite for URL-based ONNX model caching with fetch progress tracking
- **`vite.config.ts`**: Replace `@huggingface/transformers` exclusion with `onnxruntime-web`
- Remove `audioSegment.ts` (loudest segment extraction no longer needed — EffNet processes full mel spectrogram via patches)
- Remove `@huggingface/transformers` dependency; add `essentia.js` and `onnxruntime-web`

## Test plan

- [x] All 786 existing tests pass
- [x] ESLint clean
- [x] Production build succeeds
- [ ] Manual test: upload an audio file and verify the instrument icon is assigned correctly
- [ ] Manual test: verify model download progress indicator works on first load
- [ ] Manual test: verify cached model loads instantly on subsequent sessions

Closes #268

https://claude.ai/code/session_01DN5BnXjf3FfBuaPUFumaqm

### CODE REVIEW

**Comment by vsandvold** [comment]
## Fix for CORS failure on upload

The instrument classification from this PR fails on audio file upload because `essentia.upf.edu` doesn't serve `Access-Control-Allow-Origin` headers. Both the Web Worker and main-thread fallback get CORS errors when fetching the ONNX models.

### What was done

1. **CORS fix** — Proxy model requests through same origin via relative `/models/...` paths:
   - `vite.config.ts`: dev server proxy `/models` → `essentia.upf.edu`
   - `netlify.toml`: production redirect `/models/*` → `essentia.upf.edu/models/:splat`
   - Updated model URLs in `InstrumentClassificationService.ts` and `classification.worker.ts`

2. **Worker onerror fix** — The `onerror` handler set `workerFailed=true` before rejecting pending requests, preventing the `classify` catch block from trying the main-thread fallback. Removed the premature assignment.

3. **Tests** — e2e test (`classification.spec.ts`) verifying no CORS errors on upload; unit test verifying worker crash triggers main-thread fallback.

See PR #270.

### Recommended next steps

1. Verify model downloads work on the deployed Netlify site (the redirect proxy needs to be tested with the actual `essentia.upf.edu` server)
2. Complete the unchecked manual test items from this PR (instrument icon accuracy, download progress indicator, cached model reload)
